MoreLinq/Interleave.cs

```diff
@@ -45,125 +43,63 @@ public static partial class MoreEnumerable
-        }
-
-        /// <summary>
-        /// Interleaves the elements of two or more sequences into a single sequence, applying the specified strategy when sequences are of unequal length
-        /// </summary>
-        /// <remarks>
-        /// Interleave combines sequences by visiting each in turn, and returning the first element of each, followed
-        /// by the second, then the third, and so on. So, for example:<br/>
-        /// <code><![CDATA[
-        /// {1,1,1}.Interleave( {2,2,2}, {3,3,3} ) => { 1,2,3,1,2,3,1,2,3 }
-        /// ]]></code>
-        /// This operator behaves in a deferred and streaming manner.<br/>
-        /// When sequences are of unequal length, this method will use the imbalance strategy specified to
-        /// decide how to continue interleaving the remaining sequences. See <see cref="ImbalancedInterleaveStrategy"/>
-        /// for more information.<br/>
-        /// The sequences are interleaved in the order that they appear in the <paramref name="otherSequences"/>
-        /// collection, with <paramref name="sequence"/> as the first sequence.
-        /// </remarks>
-        /// <typeparam name="T">The type of the elements of the source sequences</typeparam>
-        /// <param name="sequence">The first sequence in the interleave group</param>
-        /// <param name="imbalanceStrategy">Defines the behavior of the operator when sequences are of unequal length</param>
-        /// <param name="otherSequences">The other sequences in the interleave group</param>
-        /// <returns>A sequence of interleaved elements from all of the source sequences</returns>
-
-        static IEnumerable<T> Interleave<T>(this IEnumerable<T> sequence, ImbalancedInterleaveStrategy imbalanceStrategy, params IEnumerable<T>[] otherSequences)
         {
             if (sequence == null) throw new ArgumentNullException(nameof(sequence));
             if (otherSequences == null) throw new ArgumentNullException(nameof(otherSequences));
-            if (otherSequences.Any(s => s == null))
-                throw new ArgumentNullException(nameof(otherSequences), "One or more sequences passed to Interleave was null.");
 
-            return _(); IEnumerable<T> _()
-            {
-                var sequences = new[] { sequence }.Concat(otherSequences);
+            return InterleaveSkip(otherSequences.Prepend(sequence));
+        }
 
-                // produce an iterator collection for all IEnumerable<T> instancess passed to us
-                var iterators = sequences.Select(e => e.GetEnumerator()).Acquire();
-                List<IEnumerator<T>> iteratorList = null;
+        private static IEnumerable<T> InterleaveSkip<T>(this IEnumerable<IEnumerable<T>> sequences)
+        {
+            var enumerators = new LinkedList<IEnumerator<T>>();
 
-                try
+            try
+            {
+                // First pass. create enumerators.
+                foreach (var sequence in sequences)
```

Comment:
We have [`Acquire`](https://morelinq.github.io/3.2/ref/api/html/M_MoreLinq_MoreEnumerable_Acquire__1.htm) for this so I would take advantage of it here rather than roll out a duplicate implementation that needs separate testing.

Reply:
FWIU, `Acquire` do not return a `IDisposable` so `Interleave` keep the responsibility of disposing the enumerators when the result sequence enumerator is disposed.
Achieve this with `using` is not feasible since the number of `using` depend on the sequence length.
The only way I know to achieve this in a `yield` context is to use some `try {} finally {}` blocks. Because the `finally` block is called when the enumerator is disposed.


`DisposableGroup<T>` looks to achieve this, but `DisposableGroup<T>` (and `Acquire`) return an array where I use a `LinkedList`.

I have chosen to use a `LinkedList` since `Remove` operation is O(1) there.
If we use a list or an array, where the the `Remove` operation is O(N), enumerating N sequences of 1 element will coast O(N²).
I know that it is an uncommon use case (it may be less uncommon with #697), anyway this implementation manage it.

Reply:
> FWIU, `Acquire` do not return a `IDisposable`

It would return an array of enumerators that are disposable. Question is where there is something to be gained out of not calling `GetEnumerator()` on other sequences until you get to them, but that will only optimize the case of someone eventually taking fewer elements than sequences!

Reply:
`Acquire` will not remove the necessity of a `try {} finally {}` pattern, and then is just an alias of `ToArray()`

Reply:
> `Acquire` will not remove the necessity of a `try {} finally {}` (with disposal of the enumerators in the finally block), and then is just an alias of `ToArray()`.

`Acquire` _itself_ does not need to be inside a try-finally block and so it's not the same as `ToArray()`. It doesn't mean it prevents you from having one. It is, for example, perfectly safe to return the array from `Acquire` as the return value of a function:

```c#
static Stream[] OpenFiles(params string[] paths) =>
    paths.Select(p => File.OpenRead(p)).Acquire();
```

Of course, the disposal becomes the responsibility of the caller and will need a try-catch. It does prevent that.

> * If `Acquire` is used and if an error occur while _acquiring_ (the only purpose of `Acquire` is to manage this case), elements will be disposed twice, which is useless.

It is useless but it's not a problem since it is a requirement for `IDisposable.Dispose` to be idempotent ([per docs](https://docs.microsoft.com/en-us/dotnet/api/system.idisposable.dispose#remarks)):

> If an object's [Dispose](https://docs.microsoft.com/en-us/dotnet/api/system.idisposable.dispose) method is called more than once, the object must ignore all calls after the first one. 

> I have chosen to use a `LinkedList` since `Remove` operation is O(1) there.

That's a fine choice but I am not debating the most optimal solution first. My point with `Acquire` was that it's carefully designed and tested and so might be best to keep using it instead of putting everything into question. This PR was about fixing one problem but by trying to be cleverly optimal at the same time, it may introduce others bugs (and I believe it does)! This means it takes longer to review as the scope of change expands. Instead, I find that it helps to make it work, make it right and then make it fast.

Reply:
I just push a simple (I hope) implementation of Interleave based on the array returned by `Acquire`.
The `LinkedList` implementation is in #726

Reply:
@atifaziz with the implementation using `Acquire` in this PR (currently at a7f2c6f4), calls to `GetEnumerator` are made eagerly and the code below throw:

```C#
var sequenceA = Enumerable.Range(1, 1);
var sequenceB = new BreakingSequence<int>();
sequenceA.Interleave(sequenceB).Take(1).Consume();
```

I don't know it it's acceptable